### PR TITLE
ISPN-12065 Add the anchored-keys module to the server

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.redhat.com/browse/ISPN-12065

@tristantarrant I had to set the run mode to CONTAINER ... for some reason with the default I get the following errors on the server:

```java
org.infinispan.client.hotrod.exceptions.HotRodClientException:Request for messageId=10 returned server error (status=0x85): java.lang.ClassCastException: class org.infinispan.remoting.transport.jgroups.JGroupsAddress cannot be cast to class [B (org.infinispan.remoting.transport.jgroups.JGroupsAddress is in unnamed module of loader 'app'; [B is in module java.base of loader 'bootstrap')
```

### CODE REVIEW

**Comment by tristantarrant** [comment]
@ryanemerson can you add a feature flag

**Comment by ryanemerson** [comment]
Done. The feature can be enabled/disabled via `org.infinispan.feature.anchored-keys`

File: anchored-keys/src/main/java/org/infinispan/anchored/AnchoredKeysModuleLifecycle.java [hunk 1]
```diff
@@ -43,14 +49,19 @@ public void cacheManagerStarting(GlobalComponentRegistry gcr, GlobalConfiguratio
...
-            configuration.module(AnchoredKeysConfiguration.class);
+      AnchoredKeysConfiguration anchoredKeysConfiguration = configuration.module(AnchoredKeysConfiguration.class);
       if (anchoredKeysConfiguration == null || !anchoredKeysConfiguration.enabled())
          return;
 
       assert configuration.clustering().cacheMode().isReplicated();
       assert !configuration.clustering().stateTransfer().awaitInitialTransfer();
 
+      ClassLoader classLoader = cr.getGlobalComponentRegistry().getGlobalConfiguration().classLoader();
+      Features features = new Features(classLoader);
```
Comment (pruivo) [inline anchored-keys/src/main/java/org/infinispan/anchored/AnchoredKeysModuleLifecycle.java:60]:
```suggestion
      Features features = globalConfiguration.features();
```

File: anchored-keys/src/main/java/org/infinispan/anchored/AnchoredKeysModuleLifecycle.java [hunk 2]
```diff
@@ -43,14 +49,19 @@ public void cacheManagerStarting(GlobalComponentRegistry gcr, GlobalConfiguratio
...
+      if (!features.isAvailable(ANCHORED_KEYS_FEATURE)) {
+         throw LogFactory.getLog(MethodHandles.lookup().lookupClass()).featureDisabled(ANCHORED_KEYS_FEATURE);
```
Comment (pruivo) [inline anchored-keys/src/main/java/org/infinispan/anchored/AnchoredKeysModuleLifecycle.java:62]:
```suggestion
         throw Log.CONFIG.featureDisabled(ANCHORED_KEYS_FEATURE);
```

File: anchored-keys/src/main/java/org/infinispan/anchored/AnchoredKeysModuleLifecycle.java [hunk 3]
```diff
@@ -28,6 +32,8 @@
 @InfinispanModule(name = "anchored-keys", requiredModules = "core")
 public final class AnchoredKeysModuleLifecycle implements ModuleLifecycle, DynamicModuleMetadataProvider {
 
+   public static final String ANCHORED_KEYS_FEATURE = "anchored-keys";
+
    private GlobalComponentRegistry gcr;
```
Comment (pruivo) [inline anchored-keys/src/main/java/org/infinispan/anchored/AnchoredKeysModuleLifecycle.java:37]: nitpick/offtopic: unused. just remove it.

File: anchored-keys/src/main/java/org/infinispan/anchored/AnchoredKeysModuleLifecycle.java (hunk 2 above)
Comment (pruivo) [reply anchored-keys/src/main/java/org/infinispan/anchored/AnchoredKeysModuleLifecycle.java:62]: ps. method in commons logging: `org.infinispan.commons.logging.Log;`

File: anchored-keys/src/main/java/org/infinispan/anchored/AnchoredKeysModuleLifecycle.java (hunk 1 above)
Comment (ryanemerson) [reply anchored-keys/src/main/java/org/infinispan/anchored/AnchoredKeysModuleLifecycle.java:60]: :+1: I didn't know about that method

**Comment by ryanemerson** [comment]
Updated, thanks for the reviews

**Comment by tristantarrant** [comment]
Merged

**Pull request merged**